### PR TITLE
 Slack Engine Access Groups

### PR DESCRIPTION
**Pull request by garethgreenaway**
Finally getting around to adding a PR for something I started working on at the last Salt conference.  

### What does this PR do?
Adding in the ability to include access groups for the Slack engine, then specify what commands those groups have access to run.

### What issues does this PR fix or reference?
#35267

### Previous Behavior
All Slack users had access to all defined Salt commands.

### New Behavior
Command access can be broken up into groups.

### Tests written?

No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

### CODE REVIEW

**Comment by garethgreenaway** [comment]
@mchugh19 Appreciate a quick glance at this PR since you've made some additions to the Slack engine recently.

**Comment by mchugh19** [comment]
Haven't run it, but at a glance it looks solid. Should we move the config parsing to a _utils function to keep things in sync between this, the hipchat engine, and any future chatops engines?

**Comment by cachedout** [comment]
@garethgreenaway Did you see the comment above? I want to make sure you get a chance to make any needed changes before we merge this. Thanks! :]

**Comment by garethgreenaway** [comment]
@cachedout Yup!  It's a good idea but can probably be made in a future PR.

**Comment by cachedout** [comment]
@garethgreenaway Thanks for the response. I'll take one last look at this then and hopefully get it in. :]

File: salt/engines/slack.py [hunk 1]
```diff
@@ -82,10 +105,11 @@ def _get_users(token):
 
 def start(token,
           aliases=None,
-          valid_users=None,
-          valid_commands=None,
+          valid_users=[],
```
Comment (cachedout) [inline salt/engines/slack.py]: We don't allow mutable values as kwarg defaults. Here's a good explanation as to why: http://reinout.vanrees.org/weblog/2012/04/18/default-parameters.html

Instead, our typical approach is to set `None` as the default and then switch it over inside the function implementation to a list. Let me know if you have questions.

File: salt/engines/slack.py [hunk 2]
```diff
@@ -82,10 +105,11 @@ def _get_users(token):
...
+          valid_commands=[],
```
Comment (cachedout) [inline salt/engines/slack.py]: Same issue here with the default list.

**Review by cachedout — CHANGES_REQUESTED** [review]

**Pull request merged**